tools/src/main/scala/org/apache/spark/sql/rapids/tool/qualification/RunningQualificationEventProcessor.scala

```diff
@@ -0,0 +1,254 @@
+import org.apache.spark.sql.execution.ui.{SparkListenerSQLExecutionEnd, SparkListenerSQLExecutionStart}
+
+/**
+ * This is a Spark Listener that can be used to determine if the SQL queries in
+ * the running application are a good fit to try with the Rapids Accelerator for Spark.
+ *
+ * This only supports output on a per sql query basis. It supports writing to
+ * local filesystem and most distribute filesystems and blob stores supported
+ * by Hadoop.
+ *
+ * It can be run in your Spark application by installing it as a listener:
+ * set spark.extraListeners to
+ * org.apache.spark.sql.rapids.tool.qualification.RunningQualificationEventProcessor
+ * and including the tools jar rapids-4-spark-tools_2.12-<version>.jar when you start
+ * the Spark application.
+ *
+ * The user should specify the output directory if they want the output to go to separate
+ * files, otherwise it will go to the Spark driver log:
+ *  - spark.rapids.qualification.outputDir
+ *
+ * By default, this will output results for 10 SQL queries per file and will
+ * keep 100 files. This behavior is because many blob stores don't show files until
+ * they are full written so you wouldn't be able to see the results for a running
+ * application until it finishes the number of SQL queries per file. This behavior
+ * can be configured with the following configs.
+ *  - spark.rapids.qualification.output.numSQLQueriesPerFile - default 10
+ *  - spark.rapids.qualification.output.maxNumFiles - default 100
+ *
+ * @param sparkConf Spark Configuration used to get configs used by this listener
+ */
+class RunningQualificationEventProcessor(sparkConf: SparkConf) extends SparkListener with Logging {
+
+  private val qualApp = new RunningQualificationApp(true)
+  private val listener = qualApp.getEventListener
+  private val isInited = new AtomicBoolean(false)
+  private val maxSQLQueriesPerFile: Long =
+    sparkConf.get("spark.rapids.qualification.output.numSQLQueriesPerFile", "10").toLong
+  private val maxNumFiles: Int =
+    sparkConf.get("spark.rapids.qualification.output.maxNumFiles", "100").toInt
+  private val outputFileFromConfig = sparkConf.get("spark.rapids.qualification.outputDir", "")
+  private lazy val appName = qualApp.appInfo.map(_.appName).getOrElse("")
+  private var fileWriter: Option[RunningQualOutputWriter] = None
+  private var currentFileNum = 0
+  private var currentSQLQueriesWritten = 0
+  private val filesWritten = Array.fill[Seq[Path]](maxNumFiles)(Seq.empty)
+  private lazy val txtHeader = qualApp.getPerSqlTextHeader
+
+  class QualCleanerListener extends SparkListener with CleanerListener with Logging {
+    def accumCleaned(accId: Long): Unit = {
+      // remove the accums when no longer referenced
+      qualApp.cleanupAccumId(accId)
+    }
+    def rddCleaned(rddId: Int): Unit = {}
+    def shuffleCleaned(shuffleId: Int): Unit = {}
+    def broadcastCleaned(broadcastId: Long): Unit = {}
+    def checkpointCleaned(rddId: Long): Unit = {}
```

Comment:
Just want to make sure I understand it. Is `CleanerListener` an interface due to which we are declaring these methods without any implementation(as we don't care about rdd, broadcast etc) ?

Reply:
yeah https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/ContextCleaner.scala#L313. CleanerListener is called when Spark ContextCleaner ends up cleaning up unused stuff.. The particualr issue with qualifiication tool is when to clean up accumulators that are no longer used, we can't just clean when sql query is finished because stages can be reused between sql queries and accumulables as well.